tests/unit/test_utils.py

```diff
@@ -94,3 +96,146 @@ def foo(bar):
 """
         in captured.err
     )
+
+
+@pytest.mark.parametrize(
+    "principal, resolve_id, expect_error",
+    (
+        # -- Success Cases --
+        # Username which resolves to a UUID
+        ("foo@globus.org", True, False),
+        # Identity URN
+        ("urn:globus:auth:identity:64831427-e501-420b-93da-9e4efee3dd51", False, False),
+        # Group URN
+        ("urn:globus:groups:id:449718d5-32b2-48ba-bd46-e045deab5430", False, False),
+        # Plain UUID (Assume it's an identity)
+        ("64831427-e501-420b-93da-9e4efee3dd51", True, False),
+        # -- Error Cases --
+        # Username which doesn't resolve to a UUID
+        ("bar@globus.org", False, True),
+    ),
+)
+def test_resolve_principal_urn__when_principal_type_is_omitted(
+    principal, resolve_id, expect_error
+):
+    """
+    Verifies success & error cases for principal resolution when principal_type is
+      omitted (i.e. None).
+    """
+    auth_client = unittest.mock.Mock(spec=CustomAuthClient)
+    resolved_id = "64831427-e501-420b-93da-9e4efee3dd51" if resolve_id else None
+    auth_client.maybe_lookup_identity_id.return_value = resolved_id
+
+    try:
+        urn = resolve_principal_urn(
+            auth_client=auth_client,
+            principal=principal,
+            principal_type=None,
+        )
+        assert not expect_error
+        assert (
+            urn == "urn:globus:auth:identity:64831427-e501-420b-93da-9e4efee3dd51"
+            or urn == "urn:globus:groups:id:449718d5-32b2-48ba-bd46-e045deab5430"
+        )
+    except click.UsageError as e:
+        assert expect_error
+        assert e.message == (
+            f"'--principal-type' was unspecified and '{principal}' was not resolvable "
+            f"to a globus identity."
+        )
```

Comment:
I'm finding this pattern in the tests, of trying, then having assertions in the `try`-block, and otherwise branching to the `except`-block, to be hard to read. I think the main thing is that the part which is common to both paths (`resolve_principal_urn(...)`) is grouped with assertions that are only relevant in half of the cases.

So, with that motivation in mind, I'm going to suggest this change, to move the `assert`s down into an else block:
```suggestion
    except click.UsageError as e:
        assert expect_error
        assert e.message == (
            f"'--principal-type' was unspecified and '{principal}' was not resolvable "
            f"to a globus identity."
        )
    else:
        assert not expect_error
        assert (
            urn == "urn:globus:auth:identity:64831427-e501-420b-93da-9e4efee3dd51"
            or urn == "urn:globus:groups:id:449718d5-32b2-48ba-bd46-e045deab5430"
        )
```

How would you feel about such a change across this whole test file?

I'm trying to respond to the fact that I was confused a bit on my first read, but I'll also note that I _did_ adjust and was able to understand all of the tests, so it's not critical.

Reply:
I find the proposed structure a lot harder to read actually. It's pretty really unintituitve to me that I'd look for the good path after the error declaration in a different nested block (a reason I'd only use try/except/else in very extreme contexts).

Would explicit comments chunking out the sections make the reading experience better for you with the existing structure? I want to be very sympathetic to the plight of a code reader.

Something like 
```python
    #  === Setup ===
    auth_client = unittest.mock.Mock(spec=CustomAuthClient)
    resolved_id = "64831427-e501-420b-93da-9e4efee3dd51" if resolve_id else None
    auth_client.maybe_lookup_identity_id.return_value = resolved_id

    try:
        # === Act ===
        urn = resolve_principal_urn(
            auth_client=auth_client,
            principal=principal,
            principal_type=None,
        )

       # === Verify Expected Success ===
       assert not expect_error
       assert (
            urn == "urn:globus:auth:identity:64831427-e501-420b-93da-9e4efee3dd51"
            or urn == "urn:globus:groups:id:449718d5-32b2-48ba-bd46-e045deab5430"
       )
    except click.UsageError as e:
        # === Verify Expected Failure ===
        assert expect_error
        assert e.message == (
            f"'--principal-type' was unspecified and '{principal}' was not resolvable "
            f"to a globus identity."
        )
```

Reply:
I think the comments help a little but don't really resolve the readability quirk -- TBH, if that's the only option, I'd just as soon leave it all alone.

For me, the disorientation comes from the fact that `try-except` is actually encoding branching here -- which is not what it does as flow-control, but is what it's doing vis-a-vis test logic. `try-except-else` is a branching construct, so it logically fits what the test is doing much better (whether or not you find it readable is a separate matter from how the language defines these constructs).

Perhaps this is over-valuing terseness and avoiding repetition which would be an improvement to readability. The following reconstruction seems like the most straightforward way to encode this:
```python
    if expect_error:
        with pytest.raises(
            click.UsageError, 
            match=(
                f"'--principal-type' was unspecified and '{principal}' was not resolvable "
                "to a globus identity."
            ),
        ):
            resolve_principal_urn(
                auth_client=auth_client,
                principal=principal,
                principal_type=None,
            )
    else:
        urn = resolve_principal_urn(
            auth_client=auth_client,
            principal=principal,
            principal_type=None,
        )
        assert (
             urn == "urn:globus:auth:identity:64831427-e501-420b-93da-9e4efee3dd51"
             or urn == "urn:globus:groups:id:449718d5-32b2-48ba-bd46-e045deab5430"
        )
```

---

> It's pretty unintituitve to me to look for the happy path in a different nested block after the error declaration (a reason I'd only use try/except/else in very extreme contexts).

We could probably have a much richer/fuller conversation about `try-except-else-finally` sometime, but I think we should avoid drawing conclusions from test code regarding these constructs which we'd carry into production code.

If this were _not_ a test, I would be much more vigorously requesting that we scope down the `try` block to only the action which has some known chance of failing -- the considerations are different for the different contexts.

Reply:
I don't think we're gonna agree on this one.

I'm fairly familiar with the branching logic of `try-except-else-finally`.
In this case, I think that the most formally accurate approach (`try-except-else`) is both:
 1. quite a bit more unintuitive to read
 2. only more formally accurate in an trivially unimpactful way (if of the two assertions somehow raises a click UsageError we'd apply error-expectation semantics which would fail with a slightly more inaccurate error path)

---

I think we're spending too long on this part of the PR though so I'll update to your `try/except/else` semantics as that's the most palatable of the options you're happy with for me & it's minor enough that I'd rather not spend more cycles on it.

New rev coming.

Reply:
All a very fair assessment. We might be too far apart on this to agree quickly, and it's too minor of a topic for us to devote too much time to it.

I would have also been okay with us agreeing to close this as "we don't agree and therefore we'll leave it as it was", but I appreciate you biasing towards my input as a reviewer.
Readability is very subjective, so it's not something I like to push too hard, but I am always pretty eager to discuss it. Rather than continuing to belabor this topic though... Approval incoming. :wink: